contracts/DxMarketMaker.sol

```diff
@@ -0,0 +1,501 @@
+        );
+        dx.postBuyOrder(sellToken, buyToken, auctionIndex, buyTokenAmount);
+        emit BoughtInAuction(
+            sellToken,
+            buyToken,
+            auctionIndex,
+            buyTokenAmount,
+            willClearAuction
+        );
+        return true;
+    }
+
+    function willAmountClearAuction(
+        address sellToken,
+        address buyToken,
+        uint auctionIndex,
+        uint amount
+    )
+        public
+        view
+        returns (bool)
+    {
+        // TODO: add similar requires here and in other places?
+        // R1: auction must not have cleared
+        // require(closingPrices[sellToken][buyToken][auctionIndex].den == 0);
+
+        // uint auctionStart = getAuctionStart(sellToken, buyToken);
+
+        // R2
+        // require(auctionStart <= now);
+
+        // R4
+        // require(auctionIndex == getAuctionIndex(sellToken, buyToken));
+
+        // R5: auction must not be in waiting period
+        // require(auctionStart > AUCTION_START_WAITING_FOR_FUNDING);
+
+        uint buyVolume = dx.buyVolumes(sellToken, buyToken);
+
+        // R7
+        // require(add(buyVolume, amount) < 10 ** 30);
+
+        // Overbuy is when a part of a buy order clears an auction
+        // In that case we only process the part before the overbuy
+        // To calculate overbuy, we first get current price
+        uint sellVolume = dx.sellVolumesCurrent(sellToken, buyToken);
+
+        uint num;
+        uint den;
+        (num, den) = dx.getCurrentAuctionPrice(sellToken, buyToken, auctionIndex);
+        // 10^30 * 10^37 = 10^67
+        uint outstandingVolume = atleastZero(int(mul(sellVolume, num) / den - buyVolume));
+        return amount >= outstandingVolume;
+    }
+
+    function magic(
+        address sellToken,
+        address buyToken
+    )
+        public
+        returns (bool)
+    {
+        AuctionState state = getAuctionState(sellToken, buyToken);
+
+        if (state == AuctionState.AUCTION_TRIGGERED_WAITING) {
+            return false;
+        }
+
+        if (state == AuctionState.NO_AUCTION_TRIGGERED) {
+            claimAuctionTokens(sellToken, buyToken);
+            triggerAuction(sellToken, buyToken);
+            return true;
+        }
+
+        if (state == AuctionState.AUCTION_IN_PROGRESS) {
+            // TODO: extract into a new function
+            uint auctionIndex = dx.getAuctionIndex(sellToken, buyToken);
+            uint amount = calculateAuctionBuyTokens(
+                sellToken,
+                buyToken,
+                auctionIndex,
+                address(this)
+            );
+            uint dNum;
+            uint dDen;
+            (dNum, dDen) = dx.getCurrentAuctionPrice(sellToken, buyToken, auctionIndex);
+            uint kNum;
+            uint kDen;
+            (kNum, kDen) = getKyberRate(sellToken, buyToken, amount);
+
+            // TODO: Check for overflow explicitly?
+            if (mul(dNum, kDen) > mul(kNum, dDen)) {
```

Comment:
there is a decimal issue when comparing the rate.
The short term solution is to support only 18 decimals token. But please assert it in the code.